sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java

```diff
@@ -149,6 +150,26 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
         return Channels.newChannel(
             new BZip2CompressorInputStream(Channels.newInputStream(channel)));
       }
+    },
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     */
+    ZIP {
+      @Override
+      public boolean matches(String fileName) {
+        return fileName.toLowerCase().endsWith(".zip");
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        // TODO deal with multi-entries zip
+        if (zip.getNextEntry() != null) {
+          return Channels.newChannel(zip);
+        }
+        throw new IllegalArgumentException("ZIP file doesn't contain any entry");
```

Comment:
What will the behavior be today on a multi-entry zip? Will it silently produce bad data? Fail in some way?

Please comment, and then also add a test.